c_src/membrane_ffmpeg_video_filter_plugin/text_overlay.c

```diff
@@ -0,0 +1,208 @@
+                              filtered_frame->linesize, filtered_frame->format,
+                              filtered_frame->width, filtered_frame->height,
+                              1) < 0) {
+    res = filter_result_error(env, "copy_to_payload");
+    goto exit_filter;
+  }
+  res = filter_result_ok(env, &payload_frame);
+exit_filter:
+  av_frame_free(&frame);
+  av_frame_free(&filtered_frame);
+  return res;
+}
+
+void handle_destroy_state(UnifexEnv *env, State *state) {
+  UNIFEX_UNUSED(env);
+  if (state->filter_graph != NULL) {
+    avfilter_graph_free(&state->filter_graph);
+  }
+}
+
+void create_filter_description(char *filter_descr, int len, char *text,
+                               int fontsize, int box, char *boxcolor,
+                               int border, char *fontcolor, char *fontfile,
+                               char *x, char *y) {
+
+  filter_descr += snprintf(filter_descr, len, "drawtext=text=%s", text);
+  if (fontsize != -1) {
+    filter_descr += snprintf(filter_descr, len, ":fontsize=%d", fontsize);
+  }
+  if (box != -1) {
+    filter_descr += snprintf(filter_descr, len, ":box=%d", box);
+  }
+  if (strcmp(boxcolor, "") != 0) {
+    filter_descr += snprintf(filter_descr, len, ":boxcolor=%s", boxcolor);
+  }
+  if (strcmp(fontcolor, "") != 0) {
+    filter_descr += snprintf(filter_descr, len, ":fontcolor=%s", fontcolor);
+  }
+  if (strcmp(fontfile, "") != 0) {
+    filter_descr += snprintf(filter_descr, len, ":fontfile=%s", fontfile);
+  }
+  if (border == 1) {
+    filter_descr +=
+        snprintf(filter_descr, len, ":bordercolor=DarkGray:borderw=1");
+  }
+  if (strcmp(x, "center") == 0) {
+    filter_descr += snprintf(filter_descr, len, ":x=(w-text_w)/2");
+  } else if (strcmp(x, "right") == 0) {
+    // leave 1% margin to the border
+    filter_descr += snprintf(filter_descr, len, ":x=(w-text_w)-w/100");
+  } else if (strcmp(x, "left") == 0) {
+    // leave 1% margin to the border
+    filter_descr += snprintf(filter_descr, len, ":x=w/100");
+  } else { // literal
+    filter_descr += snprintf(filter_descr, len, ":x=%s", x);
+  }
+  if (strcmp(y, "center") == 0) {
+    filter_descr += snprintf(filter_descr, len, ":y=(h-text_h)/2");
+  } else if (strcmp(y, "top") == 0) {
+    // set the same margin for width and height
+    filter_descr += snprintf(filter_descr, len, ":y=w/100");
+  } else if (strcmp(y, "bottom") == 0) {
+    filter_descr += snprintf(filter_descr, len, ":y=(h-text_h)-w/100");
+  } else { // literal
+    filter_descr += snprintf(filter_descr, len, ":y=%s", y);
+  }
+}
+int get_pixel_format(char *fmt_name) {
+  int pix_fmt = -1;
+  if (strcmp(fmt_name, "I420") == 0) {
+    pix_fmt = AV_PIX_FMT_YUV420P;
+  } else if (strcmp(fmt_name, "I422") == 0) {
+    pix_fmt = AV_PIX_FMT_YUV422P;
+  } else if (strcmp(fmt_name, "I444") == 0) {
+    pix_fmt = AV_PIX_FMT_YUV444P;
+  }
+  return pix_fmt;
+}
+UNIFEX_TERM create(UnifexEnv *env, char *text, int width, int height,
```

Comment:
Please reorder the functions to group them - the ones used for create, then filter (or apply_filter) then the desctructor